cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,231 @@
+
+### Q : Reason behind lazily updating the DNS rules in the ep_config.json file?
+
+### Discovering endpoint metadata from SDP
+
+In order to enforce L7 DNS policy, SDP needs to translate IP address to endpoint ID and identity ID. The simplest option is to reuse the same gRPC stream and implement another method to get the mappings from cilium agent. However, the actual source of truth are the bpf maps and SDP is expected to resolve the DNS queries even when agent is down, so we could read the information directly from the bpf maps. We prefer the bpf option since it does not rely on agent availability.
+
+#### Option 1a: gRPC method
+
+Get the endpoint ID and identity for a given IP address by making a gRPC call to cilium agent.
+
+##### Pros
+
+* Avoid interacting with low-level details. Simpler to maintain and stay upto date with any datapath changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Reliance on gRPC call in the hot-path
+* Does not cache the data for endpoints that never made a DNS query.
+* In an event where SDP restarts when agent's gRPC service is unavailable, all cached state is lost and SDP cannot translate IP to endpoint ID or identity.
+
+#### Option 1b: Listen to ipcache updates via grpc
+
+Similar to envoy, SDP can listen to ipcache updates and maintain a local cache of IP to endpoint ID mappings via grpc. This changes the way DNS proxy currently behaves i.e pull based model to push based model.
+
+##### Pros
+
+* Simpler to maintain and stay upto date with any new ip<>identity changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Need a mechanism for reconciling the cache in case of SDP restarts.
+* SDP will be aware of all the endpoints data, which might not be needed for DNS proxy.
+
+#### Option 2: Read from bpf maps
+
+Read mappings from bpf maps `cilium_lxc` for endpoint ID and `cilium_ipcache` for identity.
+
+##### Pros
+
+* Reduced resource utilization in agent since agent doesn't need to process rpc calls in hotpath
+* IP to endpoint ID or identity translation does not rely on agent availability.
+* DNS can be resolved even when agent is down.
+
+##### Cons
+
+* Low level interactions from components external to cilium agent.
+
+### Flag to disable built-in proxy
+
+Adding a flag in cilium agent to disable built-in dns proxy allows for de-coupling of SDP's lifecycle from agent. The current implementation of cilium agent depends on the built-in dns proxy for restoring the endpoint for DNS rules. With the flag enabled we will need to refactor to remove that dependency. Cilium agent can retrieve the DNS rules from in memory policy representation (agent already parses this information from ep_config.json on startup).
+
+#### Pros
+
+* Allows for de-coupling upgrade cycles of SDP and cilium agent.
```

Comment:
The idea behind disabling agent dnsproxy and running only SDP is that we don't need to maintain two copies of instance and if any vulnerability/issue find in dnsproxy component, we would able to upgrade SDP alone without taking down cilium agent..

Reply:
Why does upgrading SDP alone without taking down cilium-agent matter? From a DNS proxy HA perspective, you would take one component down at a time and upgrade them to mitigate the issue. In order to reconfigure Cilium to disable the built-in proxy, you would need to restart it anyway, thereby introducing a similar level of risk into the environment. Then longer term after the security event you would want to migrate back onto having Cilium in the main path for better resource utilization / efficiency.

Reply:
why do I need to maintain 2 instances if I can achieve my goal with 1 instance? Also, I don't require to take cilium agent down at all as its critical system component if I just need to upgrade SDP. Taking cilium agent would affect pod creation/deletion, policy creation/deletion during that time window

> In order to reconfigure Cilium to disable the built-in proxy, you would need to restart it anyway, thereby introducing a similar level of risk into the environment

This will happen only if customer upgrades. Also once its upgraded, later if there any patch or cve need to be fixed in dnsproxy, we are not going to restart agent but just SDP

>  Then longer term after the security event you would want to migrate back onto having Cilium in the main path for better resource utilization / efficiency.

There is definitely trade off between maintainability and performance. Thats why i'm not proposing this as default instead user can have opt-in to disable agent proxy based on their requirement.

Reply:
I still don't understand the use case. Users will have to run both the agent and SDP anyway, and I don't see any difference to maintainability if one or both components are handling DNS proxy traffic. They're also both components that will need to be maintained, upgraded, etc.

The agent will need to have DNS proxying enabled in order to restart/upgrade the SDP without downtime, or otherwise every time you upgrade SDP you would have to *also* cycle the agent to enable/disable DNS proxying in the agent. The overall goal of the CFP is high availability, meaning that if one component goes down, the other component is still available to provide service. I don't understand how that squares with the idea of reducing the number of regular running instances down to 1. If you want one DNS proxy, you can do that today by just running cilium-agent.

Now, whether or not the traffic is actually directed towards both cilium-agent and SDP can be another question - I think there are probably ways to guide that traffic towards one or both active dnsproxy instances.

Reply:
> The agent will need to have DNS proxying enabled in order to restart/upgrade the SDP without downtime,

This is not required if we do max surge upgrade for SDP which guarantees old SDP can run along with new SDP until new sdp ready for min seconds.  https://kubernetes.io/blog/2022/09/15/app-rollout-features-reach-stable/#how-use-daemonset-maxsurge

Anyway, as you said in other comments that this is not core part of this CFP. Can create a separate ticket for this specifically and can discuss there. so that dont want to deviate away from core idea (which is bootstrapping SDP as separate component)